### PR TITLE
fle-dump: a fast leader election algorithm sniffer

### PR DESCRIPTION
**Pull request by rgs1**
See README for details on usage.

Signed-off-by: Raul Gutierrez S <rgs@twitter.com>

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -154,6 +155,54 @@ Other relevant endpoints for stats are:
 * /json/auths: per-auth stats
 * /json/auths-dump: a full dump of known auths
 
+### More tools! ###
+
+Along with zk-dump and zk-stats-daemon, you can find fle-dump which allows you
+to inspect FastLeaderElection traffic (i.e.: the protocol by which ZooKeeper decides
+who will lead and also the mechanism by which the leader is subsequently discovered):
```
Comment (chchen) [inline README.md]: Ditch the "also" in "also the mechanism"

File: bin/fle-dump [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==================================================================================================
+
+from collections import deque
+import os
+from threading import Thread
+import time
+import sys
```
Comment (chchen) [inline bin/fle-dump]: I generally separate the two sorts of import statements out, but whatever works for you

File: bin/fle-dump [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+
+if os.getenv("ZKTRAFFIC_SOURCE") is not None:
+  sys.path.insert(0,  "..")
+  sys.path.insert(0, ".")
```
Comment (chchen) [inline bin/fle-dump]: Consider making this an option, rather than a env variable? I just don't like relative things in paths

File: bin/fle-dump [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+
+  def run(self):
+    while True:
+      try:
+        msg = self._queue.popleft()
+      except IndexError:
+        time.sleep(0.1)
+        continue
+
+      self._print(msg)
```
Comment (chchen) [inline bin/fle-dump]: Collapse this into:

try:
  self._print(self._queue.popleft())
except IndexError:
  time.sleep(0.1)

...?

File: zktraffic/fle/message.py [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==================================================================================================
+
+from datetime import datetime
+
+from zktraffic.base.network import BadPacket
+from zktraffic.base.util import read_long, read_number, read_string
```
Comment (chchen) [inline zktraffic/fle/message.py]: I prefer doing multiple imports this way:

from foo.bar import {
  baz,
  biz
}

At least I think it works that way. Makes diffs really easy to read.

File: zktraffic/fle/message.py [hunk 6]
```diff
@@ -0,0 +1,133 @@
...
+
+      leader, offset = read_long(data, offset)
+      zxid, offset = read_long(data, offset)
+      election_epoch, offset = read_long(data, offset)
+      peer_epoch, offset = read_long(data, offset) if len(data) > 28 else (-1, offset)
+      config = data[36:] if len(data) > 36 else ""
+
+      return Notification(src, state, leader, zxid, election_epoch, peer_epoch, config)
+
+    raise BadPacket("Unknown unknown")
```
Comment (chchen) [inline zktraffic/fle/message.py]: SO MANY MAGIC NUMBERS

File: bin/fle-dump (hunk 3 above)
Comment (rgs1) [inline bin/fle-dump]: an option would mean doing imports further down in the file. this is ugly but convenient for running/testing from source so we'll live with it, alas

File: zktraffic/fle/message.py (hunk 5 above)
Comment (rgs1) [inline zktraffic/fle/message.py]: --diff-word works too! i generally agree, but since it's < 100 cols, leaving it that way for now

**Comment by rgs1** [comment]
most comments addressed, thx @chchen!

**Pull request merged**